cpp/src/arrow/filesystem/localfs.cc

```diff
@@ -0,0 +1,387 @@
+  RETURN_NOT_OK(PlatformFilename::FromString(path, &fn));
+  if (recursive) {
+    return ::arrow::internal::CreateDirTree(fn);
+  } else {
+    return ::arrow::internal::CreateDir(fn);
+  }
+}
+
+Status LocalFileSystem::DeleteDir(const std::string& path) {
+  bool deleted = false;
+  PlatformFilename fn;
+  RETURN_NOT_OK(PlatformFilename::FromString(path, &fn));
+  RETURN_NOT_OK(::arrow::internal::DeleteDirTree(fn, &deleted));
+  if (deleted) {
+    return Status::OK();
+  } else {
+    return Status::IOError("Directory does not exist: '", path, "'");
+  }
+}
+
+Status LocalFileSystem::DeleteFile(const std::string& path) {
+  bool deleted = false;
+  PlatformFilename fn;
+  RETURN_NOT_OK(PlatformFilename::FromString(path, &fn));
+  RETURN_NOT_OK(::arrow::internal::DeleteFile(fn, &deleted));
+  if (deleted) {
+    return Status::OK();
+  } else {
+    return Status::IOError("File does not exist: '", path, "'");
+  }
+}
+
+Status LocalFileSystem::Move(const std::string& src, const std::string& dest) {
+  PlatformFilename sfn, dfn;
+  RETURN_NOT_OK(PlatformFilename::FromString(src, &sfn));
+  RETURN_NOT_OK(PlatformFilename::FromString(dest, &dfn));
+
+#ifdef _WIN32
+  if (!MoveFileExW(sfn.ToNative().c_str(), dfn.ToNative().c_str(),
+                   MOVEFILE_REPLACE_EXISTING)) {
+    return WinErrorToStatus("Failed renaming '", sfn.ToString(), "' to '", dfn.ToString(),
+                            "': ");
+  }
+#else
+  if (rename(sfn.ToNative().c_str(), dfn.ToNative().c_str()) == -1) {
+    return ErrnoToStatus("Failed renaming '", sfn.ToString(), "' to '", dfn.ToString(),
+                         "': ");
+  }
+#endif
+  return Status::OK();
+}
+
+Status LocalFileSystem::CopyFile(const std::string& src, const std::string& dest) {
+  PlatformFilename sfn, dfn;
+  RETURN_NOT_OK(PlatformFilename::FromString(src, &sfn));
+  RETURN_NOT_OK(PlatformFilename::FromString(dest, &dfn));
+  // XXX should we use fstat() to compare inodes?
+  if (sfn.ToNative() == dfn.ToNative()) {
+    return Status::OK();
+  }
+
+#ifdef _WIN32
+  if (!CopyFileW(sfn.ToNative().c_str(), dfn.ToNative().c_str(),
+                 FALSE /* bFailIfExists */)) {
+    return WinErrorToStatus("Failed copying '", sfn.ToString(), "' to '", dfn.ToString(),
+                            "': ");
+  }
+  return Status::OK();
+#else
+  std::shared_ptr<io::InputStream> is;
+  std::shared_ptr<io::OutputStream> os;
+  RETURN_NOT_OK(OpenInputStream(src, &is));
+  RETURN_NOT_OK(OpenOutputStream(dest, &os));
+  RETURN_NOT_OK(internal::CopyStream(is, os, 1024 * 1024 /* chunk_size */));
+  RETURN_NOT_OK(os->Close());
+  return is->Close();
+#endif
+}
+
+Status LocalFileSystem::OpenInputStream(const std::string& path,
+                                        std::shared_ptr<io::InputStream>* out) {
+  PlatformFilename fn;
+  int fd;
+  std::shared_ptr<io::ReadableFile> file;
+  RETURN_NOT_OK(PlatformFilename::FromString(path, &fn));
+  RETURN_NOT_OK(::arrow::internal::FileOpenReadable(fn, &fd));
```

Comment:
Is there an advantage to not let `ReadableFile::Open` do the work of calling `internal::FileOpenReadable`?

Reply:
Good question. It looks like it should actually do the same work :-)